File: CS/Form1.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace WindowsApplication1
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
		private DevExpress.XtraGrid.GridControl gridControl1;
		private DevExpress.XtraGrid.Views.Grid.GridView gridView1;
		private DevExpress.XtraGrid.Columns.GridColumn colProduct;
		private DevExpress.XtraGrid.Columns.GridColumn colInStock;
		private DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit InStockCheckEdit;
		private DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit MachineCheckEdit;
        private DevExpress.XtraGrid.Columns.GridColumn gridColumn1;
        private DevExpress.XtraGrid.Columns.GridColumn gridColumn2;
        private DevExpress.XtraGrid.Columns.GridColumn gridColumn3;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
            DevExpress.XtraGrid.GridLevelNode gridLevelNode1 = new DevExpress.XtraGrid.GridLevelNode();
            this.gridControl1 = new DevExpress.XtraGrid.GridControl();
            this.gridView1 = new DevExpress.XtraGrid.Views.Grid.GridView();
            this.colProduct = new DevExpress.XtraGrid.Columns.GridColumn();
            this.colInStock = new DevExpress.XtraGrid.Columns.GridColumn();
            this.InStockCheckEdit = new DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit();
            this.gridColumn1 = new DevExpress.XtraGrid.Columns.GridColumn();
            this.gridColumn2 = new DevExpress.XtraGrid.Columns.GridColumn();
            this.gridColumn3 = new DevExpress.XtraGrid.Columns.GridColumn();
            this.MachineCheckEdit = new DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit();
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.InStockCheckEdit)).BeginInit();
            ((System.ComponentModel.ISupportInitialize)(this.MachineCheckEdit)).BeginInit();
            this.SuspendLayout();
            // 
            // gridControl1
            // 
            this.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill;
            gridLevelNode1.RelationName = "Level1";
            this.gridControl1.LevelTree.Nodes.AddRange(new DevExpress.XtraGrid.GridLevelNode[] {
            gridLevelNode1});
            this.gridControl1.Location = new System.Drawing.Point(0, 0);
            this.gridControl1.MainView = this.gridView1;
            this.gridControl1.Name = "gridControl1";
            this.gridControl1.RepositoryItems.AddRange(new DevExpress.XtraEditors.Repository.RepositoryItem[] {
            this.InStockCheckEdit,
            this.MachineCheckEdit});
            this.gridControl1.Size = new System.Drawing.Size(610, 354);
            this.gridControl1.TabIndex = 0;
            this.gridControl1.ViewCollection.AddRange(new DevExpress.XtraGrid.Views.Base.BaseView[] {
            this.gridView1});
            // 
            // gridView1
            // 
            this.gridView1.Columns.AddRange(new DevExpress.XtraGrid.Columns.GridColumn[] {
            this.colProduct,
            this.colInStock,
            this.gridColumn1,
            this.gridColumn2,
            this.gridColumn3});
            this.gridView1.GridControl = this.gridControl1;
            this.gridView1.GroupCount = 1;
            this.gridView1.Name = "gridView1";
            this.gridView1.SortInfo.AddRange(new DevExpress.XtraGrid.Columns.GridColumnSortInfo[] {
            new DevExpress.XtraGrid.Columns.GridColumnSortInfo(this.gridColumn1, DevExpress.Data.ColumnSortOrder.Ascending)});
            this.gridView1.CustomUnboundColumnData += new DevExpress.XtraGrid.Views.Base.CustomColumnDataEventHandler(this.gridView1_CustomUnboundColumnData);
            this.gridView1.Click += new System.EventHandler(this.gridView1_Click);
            // 
            // colProduct
            // 
            this.colProduct.Caption = "Product";
            this.colProduct.FieldName = "Product";
            this.colProduct.Name = "colProduct";
            this.colProduct.Visible = true;
            this.colProduct.VisibleIndex = 0;
            // 
            // colInStock
            // 
            this.colInStock.Caption = "Selected (unbound)";
            this.colInStock.ColumnEdit = this.InStockCheckEdit;
            this.colInStock.FieldName = "UnboundChecked";
            this.colInStock.Name = "colInStock";
            this.colInStock.OptionsColumn.AllowEdit = false;
            this.colInStock.UnboundType = DevExpress.Data.UnboundColumnType.Boolean;
            this.colInStock.Visible = true;
            this.colInStock.VisibleIndex = 1;
            // 
            // InStockCheckEdit
            // 
            this.InStockCheckEdit.AutoHeight = false;
            this.InStockCheckEdit.CheckStyle = DevExpress.XtraEditors.Controls.CheckStyles.Radio;
            this.InStockCheckEdit.Name = "InStockCheckEdit";
            // 
            // gridColumn1
            // 
            this.gridColumn1.Caption = "CategoryID";
            this.gridColumn1.FieldName = "CategoryID";
            this.gridColumn1.Name = "gridColumn1";
            this.gridColumn1.Visible = true;
            this.gridColumn1.VisibleIndex = 2;
            // 
            // gridColumn2
            // 
            this.gridColumn2.Caption = "Selected (readonly)";
            this.gridColumn2.FieldName = "Selected";
            this.gridColumn2.Name = "gridColumn2";
            this.gridColumn2.OptionsColumn.ReadOnly = true;
            this.gridColumn2.Visible = true;
            this.gridColumn2.VisibleIndex = 2;
            // 
            // gridColumn3
            // 
            this.gridColumn3.Caption = "ItemDescription";
            this.gridColumn3.FieldName = "ItemDescription";
            this.gridColumn3.Name = "gridColumn3";
            this.gridColumn3.Visible = true;
            this.gridColumn3.VisibleIndex = 3;
            // 
            // MachineCheckEdit
            // 
            this.MachineCheckEdit.AutoHeight = false;
            this.MachineCheckEdit.Name = "MachineCheckEdit";
            // 
            // Form1
            // 
            this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
            this.ClientSize = new System.Drawing.Size(610, 354);
            this.Controls.Add(this.gridControl1);
            this.Name = "Form1";
            this.Text = "Form1";
            this.Load += new System.EventHandler(this.Form1_Load);
            ((System.ComponentModel.ISupportInitialize)(this.gridControl1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.gridView1)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.InStockCheckEdit)).EndInit();
            ((System.ComponentModel.ISupportInitialize)(this.MachineCheckEdit)).EndInit();
            this.ResumeLayout(false);

        }
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

        Hashtable selectedItems = new Hashtable();
		
        private void Form1_Load(object sender, System.EventArgs e)
		{
			DataTable myTable = new DataTable();
			myTable.Columns.Add("CategoryID", typeof(int));
            myTable.Columns.Add("Selected", typeof(bool));
            myTable.Columns.Add("ItemDescription", typeof(string));
            
            myTable.Rows.Add(new object[] { 1, false, "A" });
            myTable.Rows.Add(new object[] { 1, false, "B" });
            myTable.Rows.Add(new object[] { 2, false, "C" });
            myTable.Rows.Add(new object[] { 2, false, "D" });
			gridControl1.DataSource = myTable;
            gridView1.ExpandAllGroups();
		}

		private void gridView1_Click(object sender, System.EventArgs e) {
            GridView view = (GridView)sender;
			GridHitInfo hit = view.CalcHitInfo(view.GridControl.PointToClient(MousePosition));
			if(hit.InRowCell && hit.Column.FieldName == "UnboundChecked"){
                DataRow row = gridView1.GetDataRow(hit.RowHandle);
                object key = row["CategoryID"];
                DataRow prevRow = (DataRow)selectedItems[key];
                if (row == prevRow) return;
                if(prevRow != null)
                prevRow["Selected"] = false;
                view.RefreshRow(FindRowHandleByDataRow(view,prevRow));
                selectedItems[key] = view.GetDataRow(hit.RowHandle);
                view.GetDataRow(hit.RowHandle)["Selected"] = true;
                view.RefreshRow(hit.RowHandle);
			}
		}

        private int FindRowHandleByDataRow(DevExpress.XtraGrid.Views.Grid.GridView view, DataRow row)
        {
            for (int i = 0; i < view.DataRowCount; i++)
            {
                DataRow r = view.GetDataRow(i);
                if (r == row)
                    return i;
            }
            return DevExpress.XtraGrid.GridControl.InvalidRowHandle;
        }

        private void gridView1_CustomUnboundColumnData(object sender, DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs e) {
            if(e.Column.FieldName == "UnboundChecked") {
                if (e.IsGetData)
                {
                    DataRowView row = ((DataView)((GridView)sender).DataSource)[e.ListSourceRowIndex];
                    object key = row["CategoryID"];
                    DataRow selectedRow = (DataRow)selectedItems[key];
                    e.Value = (row.Row  == selectedRow);
                }
            }
        }
	}
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace WindowsApplication1
	''' <summary>
	''' Summary description for Form1.
	''' </summary>
	Public Class Form1
		Inherits System.Windows.Forms.Form
		Private gridControl1 As DevExpress.XtraGrid.GridControl
		Private WithEvents gridView1 As DevExpress.XtraGrid.Views.Grid.GridView
		Private colProduct As DevExpress.XtraGrid.Columns.GridColumn
		Private colInStock As DevExpress.XtraGrid.Columns.GridColumn
		Private InStockCheckEdit As DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit
		Private MachineCheckEdit As DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit
		Private gridColumn1 As DevExpress.XtraGrid.Columns.GridColumn
		Private gridColumn2 As DevExpress.XtraGrid.Columns.GridColumn
		Private gridColumn3 As DevExpress.XtraGrid.Columns.GridColumn
		''' <summary>
		''' Required designer variable.
		''' </summary>
		Private components As System.ComponentModel.Container = Nothing

		Public Sub New()
			'
			' Required for Windows Form Designer support
			'
			InitializeComponent()

			'
			' TODO: Add any constructor code after InitializeComponent call
			'
		End Sub

		''' <summary>
		''' Clean up any resources being used.
		''' </summary>
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				If components IsNot Nothing Then
					components.Dispose()
				End If
			End If
			MyBase.Dispose(disposing)
		End Sub

		#Region "Windows Form Designer generated code"
		''' <summary>
		''' Required method for Designer support - do not modify
		''' the contents of this method with the code editor.
		''' </summary>
		Private Sub InitializeComponent()
			Dim gridLevelNode1 As New DevExpress.XtraGrid.GridLevelNode()
			Me.gridControl1 = New DevExpress.XtraGrid.GridControl()
			Me.gridView1 = New DevExpress.XtraGrid.Views.Grid.GridView()
			Me.colProduct = New DevExpress.XtraGrid.Columns.GridColumn()
			Me.colInStock = New DevExpress.XtraGrid.Columns.GridColumn()
			Me.InStockCheckEdit = New DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit()
			Me.gridColumn1 = New DevExpress.XtraGrid.Columns.GridColumn()
			Me.gridColumn2 = New DevExpress.XtraGrid.Columns.GridColumn()
			Me.gridColumn3 = New DevExpress.XtraGrid.Columns.GridColumn()
			Me.MachineCheckEdit = New DevExpress.XtraEditors.Repository.RepositoryItemCheckEdit()
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me.InStockCheckEdit, System.ComponentModel.ISupportInitialize).BeginInit()
			CType(Me.MachineCheckEdit, System.ComponentModel.ISupportInitialize).BeginInit()
			Me.SuspendLayout()
			' 
			' gridControl1
			' 
			Me.gridControl1.Dock = System.Windows.Forms.DockStyle.Fill
			gridLevelNode1.RelationName = "Level1"
			Me.gridControl1.LevelTree.Nodes.AddRange(New DevExpress.XtraGrid.GridLevelNode() { gridLevelNode1})
			Me.gridControl1.Location = New System.Drawing.Point(0, 0)
			Me.gridControl1.MainView = Me.gridView1
			Me.gridControl1.Name = "gridControl1"
			Me.gridControl1.RepositoryItems.AddRange(New DevExpress.XtraEditors.Repository.RepositoryItem() { Me.InStockCheckEdit, Me.MachineCheckEdit})
			Me.gridControl1.Size = New System.Drawing.Size(610, 354)
			Me.gridControl1.TabIndex = 0
			Me.gridControl1.ViewCollection.AddRange(New DevExpress.XtraGrid.Views.Base.BaseView() { Me.gridView1})
			' 
			' gridView1
			' 
			Me.gridView1.Columns.AddRange(New DevExpress.XtraGrid.Columns.GridColumn() { Me.colProduct, Me.colInStock, Me.gridColumn1, Me.gridColumn2, Me.gridColumn3})
			Me.gridView1.GridControl = Me.gridControl1
			Me.gridView1.GroupCount = 1
			Me.gridView1.Name = "gridView1"
			Me.gridView1.SortInfo.AddRange(New DevExpress.XtraGrid.Columns.GridColumnSortInfo() { New DevExpress.XtraGrid.Columns.GridColumnSortInfo(Me.gridColumn1, DevExpress.Data.ColumnSortOrder.Ascending)})
'			Me.gridView1.CustomUnboundColumnData += New DevExpress.XtraGrid.Views.Base.CustomColumnDataEventHandler(Me.gridView1_CustomUnboundColumnData);
'			Me.gridView1.Click += New System.EventHandler(Me.gridView1_Click);
			' 
			' colProduct
			' 
			Me.colProduct.Caption = "Product"
			Me.colProduct.FieldName = "Product"
			Me.colProduct.Name = "colProduct"
			Me.colProduct.Visible = True
			Me.colProduct.VisibleIndex = 0
			' 
			' colInStock
			' 
			Me.colInStock.Caption = "Selected (unbound)"
			Me.colInStock.ColumnEdit = Me.InStockCheckEdit
			Me.colInStock.FieldName = "UnboundChecked"
			Me.colInStock.Name = "colInStock"
			Me.colInStock.OptionsColumn.AllowEdit = False
			Me.colInStock.UnboundType = DevExpress.Data.UnboundColumnType.Boolean
			Me.colInStock.Visible = True
			Me.colInStock.VisibleIndex = 1
			' 
			' InStockCheckEdit
			' 
			Me.InStockCheckEdit.AutoHeight = False
			Me.InStockCheckEdit.CheckStyle = DevExpress.XtraEditors.Controls.CheckStyles.Radio
			Me.InStockCheckEdit.Name = "InStockCheckEdit"
			' 
			' gridColumn1
			' 
			Me.gridColumn1.Caption = "CategoryID"
			Me.gridColumn1.FieldName = "CategoryID"
			Me.gridColumn1.Name = "gridColumn1"
			Me.gridColumn1.Visible = True
			Me.gridColumn1.VisibleIndex = 2
			' 
			' gridColumn2
			' 
			Me.gridColumn2.Caption = "Selected (readonly)"
			Me.gridColumn2.FieldName = "Selected"
			Me.gridColumn2.Name = "gridColumn2"
			Me.gridColumn2.OptionsColumn.ReadOnly = True
			Me.gridColumn2.Visible = True
			Me.gridColumn2.VisibleIndex = 2
			' 
			' gridColumn3
			' 
			Me.gridColumn3.Caption = "ItemDescription"
			Me.gridColumn3.FieldName = "ItemDescription"
			Me.gridColumn3.Name = "gridColumn3"
			Me.gridColumn3.Visible = True
			Me.gridColumn3.VisibleIndex = 3
			' 
			' MachineCheckEdit
			' 
			Me.MachineCheckEdit.AutoHeight = False
			Me.MachineCheckEdit.Name = "MachineCheckEdit"
			' 
			' Form1
			' 
			Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
			Me.ClientSize = New System.Drawing.Size(610, 354)
			Me.Controls.Add(Me.gridControl1)
			Me.Name = "Form1"
			Me.Text = "Form1"
'			Me.Load += New System.EventHandler(Me.Form1_Load);
			CType(Me.gridControl1, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me.gridView1, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me.InStockCheckEdit, System.ComponentModel.ISupportInitialize).EndInit()
			CType(Me.MachineCheckEdit, System.ComponentModel.ISupportInitialize).EndInit()
			Me.ResumeLayout(False)

		End Sub
		#End Region

		''' <summary>
		''' The main entry point for the application.
		''' </summary>
		<STAThread> _
		Shared Sub Main()
			Application.Run(New Form1())
		End Sub

		Private selectedItems As New Hashtable()

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
			Dim myTable As New DataTable()
			myTable.Columns.Add("CategoryID", GetType(Integer))
			myTable.Columns.Add("Selected", GetType(Boolean))
			myTable.Columns.Add("ItemDescription", GetType(String))

			myTable.Rows.Add(New Object() { 1, False, "A" })
			myTable.Rows.Add(New Object() { 1, False, "B" })
			myTable.Rows.Add(New Object() { 2, False, "C" })
			myTable.Rows.Add(New Object() { 2, False, "D" })
			gridControl1.DataSource = myTable
			gridView1.ExpandAllGroups()
		End Sub

		Private Sub gridView1_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles gridView1.Click
			Dim view As GridView = CType(sender, GridView)
			Dim hit As GridHitInfo = view.CalcHitInfo(view.GridControl.PointToClient(MousePosition))
			If hit.InRowCell AndAlso hit.Column.FieldName = "UnboundChecked" Then
				Dim row As DataRow = gridView1.GetDataRow(hit.RowHandle)
				Dim key As Object = row("CategoryID")
				Dim prevRow As DataRow = CType(selectedItems(key), DataRow)
				If row Is prevRow Then
					Return
				End If
				If prevRow IsNot Nothing Then
				prevRow("Selected") = False
				End If
				view.RefreshRow(FindRowHandleByDataRow(view,prevRow))
				selectedItems(key) = view.GetDataRow(hit.RowHandle)
				view.GetDataRow(hit.RowHandle)("Selected") = True
				view.RefreshRow(hit.RowHandle)
			End If
		End Sub

		Private Function FindRowHandleByDataRow(ByVal view As DevExpress.XtraGrid.Views.Grid.GridView, ByVal row As DataRow) As Integer
			For i As Integer = 0 To view.DataRowCount - 1
				Dim r As DataRow = view.GetDataRow(i)
				If r Is row Then
					Return i
				End If
			Next i
			Return DevExpress.XtraGrid.GridControl.InvalidRowHandle
		End Function

		Private Sub gridView1_CustomUnboundColumnData(ByVal sender As Object, ByVal e As DevExpress.XtraGrid.Views.Base.CustomColumnDataEventArgs) Handles gridView1.CustomUnboundColumnData
			If e.Column.FieldName = "UnboundChecked" Then
				If e.IsGetData Then
					Dim row As DataRowView = (CType((CType(sender, GridView)).DataSource, DataView))(e.ListSourceRowIndex)
					Dim key As Object = row("CategoryID")
					Dim selectedRow As DataRow = CType(selectedItems(key), DataRow)
					e.Value = (row.Row Is selectedRow)
				End If
			End If
		End Sub
	End Class
End Namespace
